docu/Concepts/TechnicalRequirements/CommunityCommunication.md

```diff
@@ -0,0 +1,460 @@
+# Community Communication
+
+This document contains the detailed descriptions of the public API of a community.
+
+## Authentication/Autorization of new Community
+
+Each public API of a community has to be authenticated and autorized before.
+
+### Variant A:
+
+This could be done by following the *OpenID Connect* protocoll. To fullfil these security requirements a separate security service has to be part of the Gradido-application.
+
+Following the link [OpenID Connect](https://www.npmjs.com/package/openid-client) there can be found a server-side OpenID relying party implementation for node.js runtime.
+
+The authentication of communities base on the community-attributes *key* and *URL*, which where exchanged during the *federation process* before. In concequence a community that hasn't execute his federation well will be unknown for other communities and can't be authenticated and autorized for further cross community API calls.
+
+### Variant B:
+
+A similar solution of authentication to variant A but **without autorization** can be done by using private and public key encryption. The *community creation* process will create a private and public key and store them internally. As the third step of the federation the *community communication* background process of the new *community-A* will be startet and a sequence of service invocations will exchange the necessary security data:
+
+![../BusinessRequirements/image/AuthenticateCommunityCommunication.png](../BusinessRequirements/image/AuthenticateCommunityCommunication.png)
+
+**1.Sequence**
+
+1. the new *community-A* encrypt the community key of the existing *community-B* with its own privat key. Then it invokes the service *authenticateCommunity* at *community-B* by sending the own community key, the encrypted community key of *community-B* and a redirect URI back to *community-A* as input data. The *community-B* will search the given community key of *community-A* in the internally stored list of communities, which are a result of the previous *federation process* collected over a different medium.
```

Comment:
```
ComA.org ---   (pubKeyA, SaltA, privKeyA(SaltA,pubKeyB)              ---> ComB.org 
              <--- (pubKeyB, SaltB, privKeyB(SaltA, SaltB, pubKeyA)  ---- 
```

Das wäre mien Vorschlag, aber ich bin kein Crypto-Experte.

Tritt ein Validierungsfehler auf, wird der Call als unauthorized markiert.

Vorteil: stateless
Nachteil: Rechenaufwendig, kann umgangen werden, wenn Salt ein Datum ist und ein solcher Salt eine Gültigkeitsdauer hat - e.g. 10min. Dann wäre aber SaltB unpraktikabel - für die Validierung auf ComB Seite wäre, dann eine Prüfung des Salt-Datums notwendig um replay Attacken zu verhindern.

Reply:
Da scheint in deinem Bild ein Henne-Ei-Problem zu sein:
Wie kommt ComA für seinen ersten Request an ComB zu dem pubKeyB? Den hat ComA zu dem Zeitpunkt doch noch gar nicht, oder?
Daher benötigt man ja einen mehr-schrittigen Handshake, um diese Keys auszutauschen...

Reply:
Kommentar ins Dokument aufgenommen